src/modules/julia.rs

```diff
@@ -0,0 +1,558 @@
+async fn print<S: Into<String>>(s: S) {
+    #[cfg(target_arch = "wasm32")]
+    {
+        write_to_xterm(s.into().as_str());
+    }
+
+    #[cfg(not(target_arch = "wasm32"))]
+    {
+        use std::io::stdout;
+        use std::io::Write;
+        print!("{}", s.into());
+        stdout().flush().unwrap();
+    }
+}
+
+#[derive(Debug)]
+pub struct Package<'a> {
+    pub name: &'a str,
+    pub id: &'a str,
+    pub versions: Vec<&'a str>,
+}
+
+pub struct Julia;
+
+#[async_trait(?Send)]
+impl Module for Julia {
+    fn name(&self) -> &'static str {
+        "julia"
+    }
+
+    fn signature(&self) -> String {
+        "julia --threads auto --project .".to_string()
+    }
+
+    async fn run(&self, appconfig: &AppConfig) {
+        let mut rng = thread_rng();
+
+        // Choose `num_packages` packages and `num_artifacts` artifacts
+        // non-repeating and in random order
+        let num_packages = rng.gen_range(10..150);
+        let num_artifacts = rng.gen_range(1..10);
+        let all_packages: Vec<&Package> = JULIA_PACKAGES_LIST
+            .choose_multiple(&mut rng, num_packages + num_artifacts)
+            .collect();
+        let (packages, artifacts) = all_packages.split_at(num_packages);
+
+        // root project of the julia session
+        let project = if rng.gen::<f32>() < 0.3f32 {
+            "@v1.7"
+        } else {
+            JULIA_PACKAGES_LIST.choose(&mut rng).unwrap().name
+        };
+
+        // julia startup
+        print_banner().await;
+        csleep(rng.gen_range(50..150)).await;
+        print_julia_prompt(false).await;
+
+        // wait user input
+        csleep(rng.gen_range(500..2500)).await;
+
+        // enter pkg mode
+        print_pkg_prompt(project, true).await;
+
+        // wait user input
+        csleep(rng.gen_range(500..1500)).await;
+
+        // type "update" and press enter
+        dprint("up", rng.gen_range(100..500)).await;
+        dprint("date", rng.gen_range(100..500)).await;
+        csleep(rng.gen_range(500..1500)).await;
+        newline().await;
+
+        // wait Pkg.update() startup time
+        csleep(rng.gen_range(200..1000)).await;
+
+        log_action("Updating", "registry at `~/.julia/registries/General.toml`").await;
+        csleep(rng.gen_range(1500..5000)).await;
+
+        log_action("Resolving", "package versions...").await;
+        csleep(rng.gen_range(500..2500)).await;
+
+        install_packages(packages).await;
+
+        csleep(rng.gen_range(250..1000)).await;
+
+        download_artifacts(artifacts).await;
+
+        csleep(rng.gen_range(250..1000)).await;
+
+        update_project_and_manifest(project).await;
+
+        csleep(rng.gen_range(10..100)).await;
+
+        report_packages(packages).await;
+
+        csleep(rng.gen_range(150..500)).await;
+
+        build_artifacts(artifacts).await;
+
+        csleep(rng.gen_range(150..500)).await;
+
+        precompile(packages).await;
```

Comment:
It might look better to group the `csleeps` either before or after the other functions here. It looks kinda weirdly spaced.

Reply:
Agreed